File: src/dotnet/Ansight.Core/HostConnectionRequest.cs
using Ansight.Pairing.Models;

namespace Ansight;

/// <summary>
/// Describes a requested host connection flow.
/// </summary>
public sealed record HostConnectionRequest(
    HostConnectionRequestKind Kind = HostConnectionRequestKind.Auto,
    PairingConfigDocument? Config = null,
    string? Payload = null,
    string? SourceDescription = null,
    string? Title = null)
{
    public static HostConnectionRequest Auto(string? sourceDescription = null)
        => new(HostConnectionRequestKind.Auto, SourceDescription: sourceDescription);

    public static HostConnectionRequest SavedConfig(string? sourceDescription = null)
        => new(HostConnectionRequestKind.SavedConfig, SourceDescription: sourceDescription);

    public static HostConnectionRequest BundledConfig(string? sourceDescription = null)
        => new(HostConnectionRequestKind.BundledConfig, SourceDescription: sourceDescription);

    public static HostConnectionRequest File(string? title = null, string? sourceDescription = null)
        => new(HostConnectionRequestKind.File, SourceDescription: sourceDescription, Title: title);

    public static HostConnectionRequest QrCode(string? title = null, string? sourceDescription = null)
        => new(HostConnectionRequestKind.QrCode, SourceDescription: sourceDescription, Title: title);

    public static HostConnectionRequest PayloadText(string payload, string? sourceDescription = null)
        => new(HostConnectionRequestKind.Payload, Payload: payload, SourceDescription: sourceDescription);

    public static HostConnectionRequest ConfigValue(PairingConfigDocument config, string? sourceDescription = null)
        => new(HostConnectionRequestKind.Config, Config: config, SourceDescription: sourceDescription);
}

File: src/ios/Sources/AnsightKit/HostConnectionRequest.swift
import Foundation

public struct HostConnectionRequest: Sendable {
    public let kind: HostConnectionRequestKind
    public let config: PairingConfigDocument?
    public let payload: String?
    public let sourceDescription: String?
    public let title: String?
    public let clientName: String?

    public init(
        kind: HostConnectionRequestKind = .auto,
        config: PairingConfigDocument? = nil,
        payload: String? = nil,
        sourceDescription: String? = nil,
        title: String? = nil,
        clientName: String? = nil
    ) {
        self.kind = kind
        self.config = config
        self.payload = payload
        self.sourceDescription = sourceDescription
        self.title = title
        self.clientName = clientName
    }

    public static func auto(clientName: String? = nil, sourceDescription: String? = nil) -> HostConnectionRequest {
        HostConnectionRequest(kind: .auto, sourceDescription: sourceDescription, clientName: clientName)
    }

    public static func savedConfig(clientName: String? = nil, sourceDescription: String? = nil) -> HostConnectionRequest {
        HostConnectionRequest(kind: .savedConfig, sourceDescription: sourceDescription, clientName: clientName)
    }

    public static func bundledConfig(clientName: String? = nil, sourceDescription: String? = nil) -> HostConnectionRequest {
        HostConnectionRequest(kind: .bundledConfig, sourceDescription: sourceDescription, clientName: clientName)
    }

    public static func file(title: String? = nil, clientName: String? = nil, sourceDescription: String? = nil) -> HostConnectionRequest {
        HostConnectionRequest(kind: .file, sourceDescription: sourceDescription, title: title, clientName: clientName)
    }

    public static func qrCode(title: String? = nil, clientName: String? = nil, sourceDescription: String? = nil) -> HostConnectionRequest {
        HostConnectionRequest(kind: .qrCode, sourceDescription: sourceDescription, title: title, clientName: clientName)
    }

    public static func payloadText(_ payload: String, clientName: String? = nil, sourceDescription: String? = nil) -> HostConnectionRequest {
        HostConnectionRequest(kind: .payload, payload: payload, sourceDescription: sourceDescription, clientName: clientName)
    }

    public static func configValue(_ config: PairingConfigDocument, clientName: String? = nil, sourceDescription: String? = nil) -> HostConnectionRequest {
        HostConnectionRequest(kind: .config, config: config, sourceDescription: sourceDescription, clientName: clientName)
    }
}
